### PR TITLE
DEV-1672 automate gdcdatamodel2 generation

### PR DESCRIPTION
**Pull request by qiaouchicago**
https://jira.opensciencedatacloud.org/browse/DEV-1672

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,147 @@
+---
+# To contribute improvements to CI/CD templates, please follow the Development guide at:
+# https://docs.gitlab.com/ee/development/cicd/templates.html
+# This specific template is located at:
+# https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/ci/templates/Python.gitlab-ci.yml
```
Comment (kulgan) [inline .gitlab-ci.yml:5]: I don't think the comment about the location of this template is valid 🤔

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+.python_versions:
+  parallel:
+    matrix:
+      - REPO_PY_VERSION: ["python36"]
+
+docker_build:
+  allow_failure: true
+  rules:
+    - when: never
```
Comment (kulgan) [inline .gitlab-ci.yml:36]: What does this stage do?

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+release:
+  image: docker.osdc.io/ncigdc/python36-builder:1.4.0
+  allow_failure: true
+  rules:
+    - when: never
+
+update_reqs_for_gdcdatamodel2:
```
Comment (kulgan) [inline .gitlab-ci.yml:44]: I think this should use the `needs` option to depend on the main gdcdictionary tests completing successfully before triggering gdcdatamodel2 work. And can we rename the stage so its clearly different from the build step of gdcdictionary?

File: .gitlab-ci.yml (hunk 2 above)
Comment (qiaouchicago) [reply .gitlab-ci.yml:36]: it suppose to build a docker image for gdcdictionary.

File: .gitlab-ci.yml (hunk 3 above)
Comment (qiaouchicago) [reply .gitlab-ci.yml:44]: >I think this should use the needs option to depend on the main gdcdictionary tests completing successfully before triggering gdcdatamodel2 work
Screen Shot 2023-02-14 at 2 05 58 PM
It already does that. The later stage usually does not start until the earlier stage finish. The only exception is `code_quality` not sure why.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+include:
+  - project: nci-gdc/gitlab-templates
+    ref: master
+    file:
+      - templates/global/full.yaml
+      - templates/python/full.yaml
+
+variables:
+ BASE_CONTAINER_VERSION: "1.4.0"
+ PRE_COMMIT_DOCKER_IMAGE: docker.osdc.io/ncigdc/python3.8-builder:2.1.0
```
Comment (stilesj-uchicago) [inline .gitlab-ci.yml:19]: I think our pre-commit version needs to be 3.7 if we continue with the idea that we support the oldest supported version of python.

File: .gitlab-ci.yml (hunk 3 above)
Comment (stilesj-uchicago) [reply .gitlab-ci.yml:48]: How is secret_detection different than the pre-commit step? Is it duplicating effort?

**Review by stilesj-uchicago — APPROVED** [review]

**Comment by qiaouchicago** [comment]
>  How is secret_detection different than the pre-commit step? Is it duplicating effort?

It was the first to run. Another hook will modify the file and the .secret.baseline will be invalid. 
Moved it to last

**Comment by qiaouchicago** [comment]
> How is secret_detection different than the pre-commit step? Is it duplicating effort?

I think it is, it is gitlab's implementation. 

https://docs.gitlab.com/ee/user/application_security/secret_detection/

File: .gitlab-ci.yml (hunk 3 above)
Comment (qiaouchicago) [reply .gitlab-ci.yml:48]: I think it is, it is gitlab's implementation.

https://docs.gitlab.com/ee/user/application_security/secret_detection/

Comment (qiaouchicago) [reply .gitlab-ci.yml:48]: >And can we rename the stage so its clearly different from the build step of gdcdictionary?

Screen Shot 2023-02-14 at 3 39 29 PM

File: .gitlab-ci.yml (hunk 2 above)
Comment (kulgan) [reply .gitlab-ci.yml:40]: What happens if you remove it, not sure why one will need a docker image of the gdcdictionary

File: .gitlab-ci.yml (hunk 3 above)
Comment (kulgan) [reply .gitlab-ci.yml:48]: I think the difference is you want this stage to run only when the previous stage passes, not sure if that is handled automatically. For example, we don't want to build gdcdatamodel2 when gdcdictionary test is failing

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    - apt-get install -y python3-venv
+    - pip3 install -U setuptools_scm build twine setuptools
+    - cp -v pyproject.toml /tmp/pyproject.toml
+    - sed -i 's/local_scheme = "no-local-version"/local_scheme = "node-and-timestamp"/' pyproject.toml
+    - cat pyproject.toml
+    - python3 -m setuptools_scm
+    - python3 -m build
+    - twine check dist/*
+    - twine upload dist/*
+    - cp -v /tmp/pyproject.toml pyproject.toml
```
Comment (kulgan) [inline .gitlab-ci.yml:110]: As improvement for later, we can make use of the tox:release command defined the gdcdatamodel2's tox.ini

File: .gitlab-ci.yml (hunk 3 above)
Comment (qiaouchicago) [reply .gitlab-ci.yml:48]: Screen Shot 2023-02-14 at 4 18 46 PM

Comment (qiaouchicago) [reply .gitlab-ci.yml:48]: I think the later stage won't run if a job in the previous stage failed.

**Review by kulgan — APPROVED** [review]

File: .gitlab-ci.yml (hunk 2 above)
Comment (qiaouchicago) [reply .gitlab-ci.yml:40]: It will still run if I remove this part. It is defined in the template.

**Pull request merged**